src/permutations.jl

```diff
@@ -10,66 +10,122 @@ export
 
 
 struct Permutations{T}
-    a::T
-    t::Int
+    data::T
+    length::Int
 end
 
-Base.eltype(::Type{Permutations{T}}) where {T} = Vector{eltype(T)}
+function has_repeats(state::Vector{Int})
+    # This can be safely marked inbounds because of the type restriction in the signature.
+    # If the type restriction is ever loosened, please check safety of the `@inbounds`
+    @inbounds for outer in eachindex(state)
+        @inbounds for inner in (outer+1):lastindex(state)
+            if state[outer] == state[inner]
+                return true
+            end
+        end
+    end
+    return false
+end
+
+function increment!(state::Vector{Int}, min::Int, max::Int)
+    state[end] += 1
+    for i in reverse(eachindex(state))[firstindex(state):end-1]
+        if state[i] > max
+            state[i] = min
+            state[i-1] += 1
+        end
+    end
+end
+
+function next_permutation!(state::Vector{Int}, min::Int, max::Int)
+    while true
+        increment!(state, min, max)
+        has_repeats(state) || break
+    end
+end
+
+function Base.iterate(p::Permutations, state::Vector{Int}=fill(firstindex(p.data), p.length))
+    next_permutation!(state, firstindex(p.data), lastindex(p.data))
+    if state[firstindex(state)] > lastindex(p.data)
+        return nothing
+    end
+    [p.data[i] for i in state], state
+end
+
+function Base.length(p::Permutations)
+    length(p.data) < p.length && return 0
+    return Int(prod(big(length(p.data) - p.length + 1):big(length(p.data))))
```

Comment:
Why convert to `BigInt` then back to `Int`? If the product exceeds `typemax(Int)`, the conversion will error anyway.

Reply:
True. Would you suggest leaving it as a `BigInt`?

Reply:
That's a good question. What is the return type of `length(::Combinations)`? I'd probably opt to make it consistent with that.

Reply:
Look like `length(::Combinations)` returns an `Int64`. Should the `length` function be worried about overflow in these cases? It's not too hard create a case that overflows and returns an error with `Int64` for `combinations` and for `permutations`. But perhaps that's outside the scope of this PR.

I'm ok with having `length(::Permutations)` not converting to `BigInt` in the first place if we think that's most consistent.

Reply:
My recommendation would be to have it be consistent with `Combinations` for now and open an issue to discuss whether the `length` function should return `BigInt`. Then both can be changed at once later.